Add kinesis output support

jimmystewpot opened the pull request:
This is a PR to add an output plugin that writes to Kinesis. 
* It batches up 500 points per API PUT request to minimise the number of API requests.
* Uses the Amazon aws-sdk-go libraries for underlying Kenesis support.
* Has some supporting documentation.
* Supports multiple forms of Authentication - Environment Variables, IAMS roles or Shared Credentials.

I have attempted to keep it as simple and small as possible. The tests pass on my local system and on other amazon instances where there are provided credentials.

sparrc commented:
Thank you @jimmystewpot! It seems like your unit test is assuming there will be some authentication credentials available? Is that true?

We may not want to have a "test connect and write" for this particular output, as it is quite a bit different than the others, unless there happens to be a docker container somewhere that can Mock the kinesis write endpoint.

sparrc commented on outputs/kinesis/kinesis_output.go:
```diff
@@ -0,0 +1,155 @@
+	return errors.New("Error")
+}
+
+func formatmetric(k *KinesisOutput, point *client.Point) (string, error) {
```
you could write a unit test for this function

sparrc replied:
As opposed to the full connect and write test

jimmystewpot replied:
Updating unit tests now. will add another commit soon.

sparrc commented on outputs/kinesis/README.md:
```diff
@@ -0,0 +1,43 @@
+
+## todo
+
+* Check if the stream exists so that we have a graceful exit.
```
how will it currently exit if the stream doesn't exist?

sparrc commented on outputs/kinesis/kinesis_output.go:
```diff
@@ -0,0 +1,155 @@
+
+var sampleConfig = `
+  # Amazon REGION of kinesis endpoint.
+  # Most AWS services have a region specific endpoint this will be used by
```
I think users of this plugin know about AWS regions, you can remove these two lines (32 & 33)

jimmystewpot replied:
Updated.

sparrc commented on outputs/kinesis/kinesis_output.go:
```diff
@@ -0,0 +1,155 @@
+  # Variables.
+  region = "ap-southeast-2"
+  # Kinesis StreamName must exist prior to starting telegraf.
+  streamname = "StreamName"
```
what happens if not?

jimmystewpot replied:
Previously metrics would be lost. I have updated the functionality to hard exit with a meaningful error and updated the documentation to reflect the change.

sparrc replied:
awesome :+1:

sparrc commented on outputs/kinesis/kinesis_output.go:
```diff
@@ -0,0 +1,155 @@
+  # Kinesis StreamName must exist prior to starting telegraf.
+  streamname = "StreamName"
+  # PartitionKey as used for sharding data.
+  partitionkey = "PartitionKey"
```
is this required? Should/can it default to an empty string?

jimmystewpot replied:
Which line are you specifically calling out?

sparrc replied:
`partitionkey = "PartitionKey"`

jimmystewpot replied:
The PartitionKey is a required field in the [API](http://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecord.html). I have put it in there (rather than having a default) so that people are aware that the field is mandatory and that it should be set to something that allows them to designate a shard for any given hosts. 

Is this bad practise?

sparrc commented on outputs/kinesis/kinesis_output_test.go:
```diff
@@ -0,0 +1,41 @@
+				&credentials.SharedCredentialsProvider{},
+			}),
+	}
+	svc := kinesis.New(session.New(Config))
```
What would this be connecting to on a local system? Are you sure credentials are not required?

jimmystewpot replied:
Credentials are required, I have updated the documentation to reflect that the system must have either 
* aws keys in environment variables
* keys configured as shared credentials ~/.aws/....
* EC2 IAMS role.

sparrc replied:
Unfortunately we can't make that a requirement for running unit tests, I think we'll just have to live with this output not being fully unit tested :neutral_face:

sparrc commented on outputs/kinesis/kinesis_output.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+func init() {
+	outputs.Add("kinesis_output", func() outputs.Output {
```
One last nitpick :sweat_smile:, call the output just `kinesis` for consistency and reduce redundancy

jimmystewpot replied:
I am planning to write a kinesis monitoring plugin which is why I named it output. Is this bad practise?

sparrc replied:
@jimmystewpot, yes, it should just be named `kinesis`, in the config they will appear in two different sections `outputs.kinesis` and `plugins.kinesis`. We could also call the plugin `kinesis_consumer` which would be consistent with the `kafka` naming that we have now.

jimmystewpot replied:
I have updated the PR to include the rename.

jimmystewpot commented:
Updated the name of the plugin as per your request. I have also removed the connection unit test so that the tests can pass. 
I will keep an eye out for a mock kinesis endpoint that I can use for testing.

sparrc commented:
Awesome, thank you @jimmystewpot! I'll fixup the `go fmt` issue when I merge this
